client/service/src/chain_ops.rs

```diff
@@ -36,12 +37,116 @@ use sc_executor::{NativeExecutor, NativeExecutionDispatch};
 
 use std::{io::{Read, Write, Seek}, pin::Pin};
 use sc_client_api::BlockBackend;
+use serde_json::{de::IoRead as JsonIoRead, Deserializer, StreamDeserializer};
 
 /// Build a chain spec json
 pub fn build_spec(spec: &dyn ChainSpec, raw: bool) -> error::Result<String> {
 	Ok(spec.as_json(raw)?)
 }
 
+/// Helper enum that wraps either a binary decoder (from parity-scale-codec), or a JSON decoder (from serde_json).
+enum BlockStream<R, B> 
+where
+	R: std::io::Read + std::io::Seek,
+{
+	Binary {
+		count: u64,
+		read_block_count: u64,
+		reader: CodecIoReader<R>,
+	},
+	Json {
+		read_block_count: u64,
+		reader: StreamDeserializer<'static, JsonIoRead<R>, B>,
+	},
+}
+
+impl<R, B> Unpin for BlockStream<R, B> 
+where
+	R: Read + Seek
+{}
+
+impl<R, B> BlockStream<R, SignedBlock<B>>
+	where
+		R: Read + Seek + 'static,
+		B: BlockT + MaybeSerializeDeserialize,
+	{
+	fn new(input: R, binary: bool) -> Result<Self, Error> {
+		if binary {
+			let mut reader = CodecIoReader(input);
+			// If the file is encoded in binary format, it is expected to first specify the number
+			// of blocks that are going to be decoded. We read it and add it to our enum struct.
+			let count: u64 = Decode::decode(&mut reader)
+				.map_err(|e| format!("Error reading file: {}", e))?;
+			Ok(BlockStream::Binary {
+				count,
+				read_block_count: 0,
+				reader,
+			})
+		} else {
+			let stream_deser  = Deserializer::from_reader(input)
+				.into_iter::<SignedBlock<B>>();
+			Ok(BlockStream::Json {
+				reader: stream_deser,
+				read_block_count: 0,
+			})
+		}
+	}
+
+	/// Returns the number of blocks read thus far.
+	fn read_block_count(&self) -> u64 {
+		match self {
+			BlockStream::Binary {count: _, read_block_count, reader: _} => *read_block_count,
+			BlockStream::Json {read_block_count, reader: _} => *read_block_count,
```

Comment:
```suggestion
			BlockStream::Binary { read_block_count,  .. } => *read_block_count,
			BlockStream::Json { read_block_count, .. } => *read_block_count,
```

rust has rest patterns :wink:

Reply:
Next leveling pattern matching right here! Thanks :)